source/D2Hell/src/Archive.cpp

```diff
@@ -1 +1,167 @@
+		{
+			FOG_WriteToLogFile("Error opening file: %s", szFilePath);
+		}
+
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/**
+ * Gets the uncompressed size of a file in the MPQ archives.
+ *
+ * Static library; may be defined in multiple places than ones listed:
+ * 1.00: D2Lang.0x10004DFE
+ * 1.10: D2Lang.0x6FC145F2
+ * 1.13c: Inline
+ * 1.14c: Game.0x00514B87
+ */
+static size_t __fastcall ARCHIVE_GetFileSize(void* pArchiveHandle, void* pFileHandle, size_t* pFileSizeHigh)
+{
+	D2_ASSERT(pFileHandle != nullptr);
+
+	size_t dwFileSize = FOG_MPQFileGetSize(pFileHandle, pFileSizeHigh);
+	if (dwFileSize == 0)
+	{
+		char szArchivePath[MAX_PATH];
+#if 0  // TODO: Enable this code once the macro is set up.
+		Storm_276_GetFileName(pFileHandle, szArchivePath, 260);
+		Fog_10026(3, szArchivePath, __FILE__, __LINE__);
+#endif
+		exit(-1);
+	}
+
+	return dwFileSize;
+}
+
+/**
+ * Reads a file from the MPQ archives into a specified buffer.
+ *
+ * Static library; may be defined in multiple places than ones listed:
+ * 1.00: D2Lang.0x10004EE3
+ * 1.10: D2Lang.0x6FC14661
+ * 1.13c: D2Lang.0x6FC07C00
+ * 1.14c: Game.0x00514C61
+ */
+static void __fastcall ARCHIVE_ReadFileToBuffer(void* pArchiveHandle, void* pFileHandle, void* pBuffer, size_t dwBytesToRead)
+{
+	D2_ASSERT(pFileHandle != nullptr);
+
+	size_t dwBytes;
+	BOOL bFileReadSuccess = FOG_MPQFileRead(pFileHandle, pBuffer, dwBytesToRead, (int*)&dwBytes, 0, 0, 0);
+	if (!bFileReadSuccess)
+	{
+		char szArchivePath[MAX_PATH];
+#if 0  // TODO: Enable this code once the macro is set up.
+		Storm_276_GetFileName(pFileHandle, szArchivePath, sizeof(szArchivePath));
+		Fog_10026(3, szArchivePath, __FILE__, __LINE__);
+#endif
+		exit(-1);
+	}
+
+	D2_ASSERT(dwBytesToRead == dwBytes);
+}
+
+/**
+ * Reads a file from the MPQ archives into a specified buffer.
+ *
+ * Static library; may be defined in multiple places than ones listed:
+ * 1.00: Inline
+ * 1.10: Inline
+ * 1.13c: Inline
+ * 1.14c: Game.0x00514B60
+ */
+static void __fastcall ARCHIVE_CloseFile(void* pArchiveHandle, void* pFileHandle)
+{
+	D2_ASSERT(pFileHandle != nullptr);
+	FOG_MPQFileClose(pFileHandle);
+}
+
+/**
+ * pArchiveHandle identifier is confirmed by 1.00 (originally hArchive).
+ * 
+ * Static library; may be defined in multiple places than ones listed:
+ * 1.00: D2Lang.0x10005029
+ * 1.10: D2Lang.0x6FC14708
+ * 1.13c: D2Lang.0x6FC07EF0
+ * 1.14c: Game.0x00514D55
+ */
+void* __fastcall ARCHIVE_ReadFile(void* pArchiveHandle, const char* szFilePath, size_t* pBytesWritten, const char* szSrcPath, int nLine)
+{
+	void* pFileHandle;
+	
+	BOOL bOpenFileSucceeded = ARCHIVE_OpenFile(pArchiveHandle, szFilePath, &pFileHandle, FALSE);
+	if (!bOpenFileSucceeded)
+	{
+		return nullptr;
+	}
+
+	size_t dwFileSizeHigh;
+	size_t dwFileSize = ARCHIVE_GetFileSize(pArchiveHandle, pFileHandle, &dwFileSizeHigh);
+	// TODO: Remove the char* cast once FOG has been const corrected.
+	void* pBuffer = FOG_AllocClientMemory(dwFileSize + 800, (char*)szSrcPath, nLine, 0);
```

Comment:
If you have time to go over FOG functions declaration and if the const correctness of file paths it'd be awesome :)